daffodil-io/src/main/scala/org/apache/daffodil/io/DirectOrBufferedDataOutputStream.scala

```diff
@@ -33,30 +33,79 @@ import org.apache.daffodil.util.Maybe
 import org.apache.daffodil.util.Maybe._
 import org.apache.daffodil.util.MaybeULong
 import org.apache.daffodil.util.Misc
+import java.io.File
 
 /**
  * This simple extension just gives us a public method for access to the underlying byte array.
  * That way we don't have to make a copy just to access the bytes.
  */
-private[io] class ByteArrayOutputStreamWithGetBuf() extends java.io.ByteArrayOutputStream {
+private[io] class ByteArrayOrFileOutputStreamWithGetBuf(
+  chunkSizeInBytes: Long = 4096,
+  maxBufferSizeInBytes: Long = 2000 * (1 << 20))
+  extends java.io.ByteArrayOutputStream {
+  private var _byteArrayOS = this
+  private var _fileOS: java.io.FileOutputStream = null
+
+  var isFile: Boolean = false
+  var tempFile: File = null
+  var bytesWritten = count
+
+  def switchOS() = {
+    Assert.invariant(_fileOS == null)
+    tempFile = File.createTempFile("daffodil-", ".tmp")
+    _fileOS = new java.io.FileOutputStream(tempFile)
+    this.flush
+    this.writeTo(_fileOS)
+    bytesWritten = count
+    this.reset
+    isFile = true
+  }
```

Comment:
Is it possible that there might be a case where a user does not want to switch to a file output, but instead wants to have an exception thrown? I can maybe imagine performance cirticial systems would want to avoid this and cause a failure so they can adjust their schema. Maybe add a tunable that will throw an exception if the maxBufferSize limit for a single buffer is reached?

Reply:
I think that this may be addressed by a logging message about switching to file based output, which can suggest changing the schema to avoid this situation.